### DISCUSSION TITLE
Suggestions and change requests

### DISCUSSION THREAD

**Discussion by ilian6806** [ilian6806/stable-diffusion-webui-state#17] [Announcements]
If you have any ideas or suggestions, you can share it here :)

**Comment by rrvsh** [comment]
Could i suggest the ability to save inpainting option states as well (masked content, inpaint area etc)? Thank you!

**Reply by ilian6806** [reply]
This will be a tricky one but I will give it a try

**Reply by sahildeshpande** [reply]
Hey did you manage to implement this?

**Reply by ilian6806** [reply]
No success for now. I dont have much time...

**Comment by r7l** [comment]
It would be awesome to be able to save state to JSON file and also import these JSON files back in.

**Reply by ilian6806** [reply]
Done :)

**Reply by r7l** [reply]
Thanks allot. This is really helpful.

**Comment by enternalsaga** [comment]
can the setting state (or at least import/export button) move to the txt2img/img2img tab so I can quickly select the state i want? ref to the Config Presets extension below
![image](https://user-images.githubusercontent.com/31148570/231227593-2af642c3-8ac6-472a-bec1-b1687e32693c.png)

**Reply by ilian6806** [reply]
I added export and import buttons in quick setting buttons, next to checkpoint in top left corner. I hope this helps.

**Comment by a2ZxgdqHeSs8F51cIyuK** [comment]
Hello, could you kindly support this extension: https://github.com/mcmonkeyprojects/sd-dynamic-thresholding ?
It has a few settings I have to set up every time, having STATE remember it would be very helpful.

**Reply by ilian6806** [reply]
Unfortunately, in order to add some elements to memory, they must have some id`s in the markup. So to support some extension, the author must do this before I can memorize it. You can create an issue in their repo for this. This will not take much time.

**Comment by kantsche** [comment]
Can you post a small tutorial on how to do your own support scripts?

**Reply by ilian6806** [reply]
I dont have much time these days, but if you mean developing custom scripts, there is a wiki page [here ](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Developing-custom-scripts)

**Reply by kantsche** [reply]
I mean adding support for other extensions. If not I'll look into the controlnet one.

**Reply by ilian6806** [reply]
Yes, you can check state.ext.control-net.js for reference. The idea is:
- Create your js file state.ext.{your-extension-name}.js in /javascript
- It must contain IIFE with init method like so:
```
window.state = window.state || {};
window.state.extensions = window.state.extensions || {};
state = window.state;

state.extensions['{your-extension-name}'] = (function () {

    function init() {
    }

    return { init };
})();
```
- Add the extension to state_settings.py -> state_extensions in order to see it in Web UI settings
- Inspect the inputs you want to save to check element ids. The best case is if there is id and it contains  txt2img/img2img part. If not, the element must be selected relatively from other element (container). For more complicated elements (seelcts, multiple selects) with custom structure, comming from gradio, you can use the functions in state.utils.js.

I hope this helps. Feel free to ask anything else. If you manage to handle some extension you can always push your changes to PR :)

**Comment by ride5k** [comment]
very helpful and handy extension.  one request: can you grab the info from the "additional networks" extension that loads LORAs?  not sure if possible, but would save a lot of time selecting LORAs and weights every time.

**Comment by Faithtoken** [comment]
Would it be able to add "CLIP skip" to the information saved?

Thanks for making this add-on. It really saves me some time.

**Reply by ilian6806** [reply]
From what I see, CLIP skip is saved by the SD in the latest version

**Comment by SunGod** [comment]
It would be **glorious** if this could support "Tiled Diffusion & VAE" (aka "[multidiffusion-upscaler-for-automatic1111](https://github.com/pkuliyi2015/multidiffusion-upscaler-for-automatic1111") here:
https://github.com/pkuliyi2015/multidiffusion-upscaler-for-automatic1111?tab=readme-ov-file#%F0%9F%94%A5-tiled-vae

It's a very popular extension allowing users to produce multi-region results even with low VRAM machines, and there are a bunch of settings that get lost every time you need to restart the UI. Having State support this extension would easily save me 5 minutes every time this happens.

**Comment by thejohnd0e** [comment]
Works great, but please add the FaceSwapLab statement.

**Comment by ElenViolet** [comment]
Nice idea! It would be great if you could add the following: "Resize by" from img2img(only "Resize to" working right now), ADetailer (https://github.com/Bing-su/adetailer) settings, "Batch" with "PNG info"  tab from img2img. There is also a nuance that if the prompt was filled in not manually but through the "png info->send to img2img" method, then it will not be saved until it is edited manually.

**Comment by dougvan1701** [comment]
Any chance you could add FaceSwapLab and ReActor support?
Thank you!

**Comment by mechz01** [comment]
Does this not support saving the checkpoint being used? Is it possible to add this? thanks.
